### PR TITLE
IGVF-845-treatment-lab-award

### PR DESCRIPTION
**Pull request by khzlin**

### CODE REVIEW

File: src/igvfd/schemas/changelogs/treatment.md [hunk 1]
```diff
@@ -1,5 +1,10 @@
 ## Changelog for *`treatment.json`*
 
+### Schema version 4
+* Add `lab` and `award`.
+* Add `depletion`. It is `False` by default. If it is `True`, `amount` and `amount_units` should not be specified.
```
Comment (gabdank) [inline src/igvfd/schemas/changelogs/treatment.md:5]: I think change log should be very terse.  No need to explain defaults, or logic of other properties.

File: src/igvfd/schemas/changelogs/treatment.md [hunk 2]
```diff
@@ -1,5 +1,10 @@
...
+* Remove `amount` and `amount_units` from required properties list.
```
Comment (gabdank) [inline src/igvfd/schemas/changelogs/treatment.md:6]: remove requirement for 'amount'....

File: src/igvfd/schemas/changelogs/treatment.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 ## Changelog for *`treatment.json`*
 
+### Schema version 4
+* Add `lab` and `award`.
```
Comment (gabdank) [inline src/igvfd/schemas/changelogs/treatment.md:4]: add required lab award

File: src/igvfd/schemas/treatment.json [hunk 4]
```diff
@@ -67,6 +70,54 @@
...
+            "required": [
+                "depletion",
+                "amount"
+            ],
+            "comment": "depletion and amount must be specified if anount_untis is specified",
+            "properties": {
+                "depletion": {
+                    "const": false
+                }
+            }
```
Comment (gabdank) [inline src/igvfd/schemas/treatment.json:95]: what does the "const:false" do?

**Review by gabdank — COMMENTED** [review]
comment/question in PR

File: src/igvfd/schemas/changelogs/treatment.md (hunk 3 above)
Comment (khzlin) [reply src/igvfd/schemas/changelogs/treatment.md:4]: fixed

File: src/igvfd/schemas/treatment.json (hunk 4 above)
Comment (khzlin) [reply src/igvfd/schemas/treatment.json:95]: it restricts the value to false
https://json-schema.org/understanding-json-schema/reference/generic.html#constant-values

File: src/igvfd/schemas/treatment.json [hunk 5]
```diff
@@ -67,6 +70,54 @@
...
+        },
+        "amount_units": {
+            "required": [
+                "depletion",
+                "amount"
+            ],
+            "comment": "depletion and amount must be specified if anount_untis is specified",
```
Comment (gabdank) [inline src/igvfd/schemas/treatment.json:90]: typo "if anount_untis"

File: src/igvfd/schemas/treatment.json [hunk 6]
```diff
@@ -67,6 +70,54 @@
             ],
             "comment": "temperature must be specified if temperature_units is specified"
         },
+        "amount": {
```
Comment (gabdank) [inline src/igvfd/schemas/treatment.json:73]: I think the logic should be reversed. "if depletion is True - then amount and units should not be specified"

**Review by gabdank — CHANGES_REQUESTED** [review]

File: src/igvfd/schemas/treatment.json (hunk 5 above)
Comment (khzlin) [reply src/igvfd/schemas/treatment.json:90]: fixed

File: src/igvfd/schemas/treatment.json [hunk 7]
```diff
@@ -67,6 +70,74 @@
...
+                "properties": {
+                    "depletion": {
+                        "const": true
+                    }
+                }
+            },
+            "then": {
+                "not": {
+                    "required": [
+                        "amount_units"
```
Comment (gabdank) [inline src/igvfd/schemas/treatment.json:85]: is not now the dependency misplaced? amount_units in amount and amount in amount_units?

File: src/igvfd/schemas/treatment.json [hunk 8]
```diff
@@ -67,6 +70,74 @@
...
+                    ]
+                }
+            },
+            "else": {
+                "required": [
+                    "amount_units"
+                ]
+            }
+        },
+        "amount_units": {
```
Comment (gabdank) [inline src/igvfd/schemas/treatment.json:95]: dependency on amount_units but the restricted property inside is amount?

**Review by gabdank — CHANGES_REQUESTED** [review]
not sure i understand the dependency

**Review by gabdank — APPROVED** [review]

File: src/igvfd/types/treatment.py [hunk 9]
```diff
@@ -31,11 +31,17 @@ class Treatment(Item):
...
-        if duration is not None:
+    def title(self, treatment_term_name, amount=None, amount_units=None, duration=None, duration_units=None):
+        if duration is not None and amount is not None:
             text = 'Treated with {} {} {} for {} {}'.format(
                 amount, amount_units, treatment_term_name, duration, duration_units)
-        else:
+        elif duration is None and amount is not None:
             text = 'Treated with {} {} {}'.format(
                 amount, amount_units, treatment_term_name)
+        elif amount is None and duration is not None:
```
Comment (iyoungworth) [inline src/igvfd/types/treatment.py:41]: I think we need a separate statement to handle where depletion=True, the title should say more explicitly "depleted in { }... etc", with or without duration present

File: src/igvfd/upgrade/treatment.py [hunk 10]
```diff
@@ -23,3 +23,17 @@ def treatment_2_3(value, system):
...
+    # https://igvf.atlassian.net/browse/IGVF-845
+    notes = value.get('notes', '')
+    if 'award' not in value and 'lab' not in value:
+        value['award'] = '/awards/HG012012'
+        value['lab'] = '/labs/j-michael-cherry'
+        if 'notes' in value:
+            value['notes'] = f'{value.get("notes")}. This treatment did not have award and lab specified previously, it was upgraded to have Cherry lab and award.'
+        else:
+            value['notes'] = 'This treatment did not have award and lab specified previously, it was upgraded to have Cherry lab and award.'
+    return
```
Comment (iyoungworth) [inline src/igvfd/upgrade/treatment.py:39]: It doesn't look like depletion is assigned a default value in the schema, so I think the upgrade needs to assign a 'False' value

File: src/igvfd/tests/fixtures/schemas/treatment.py [hunk 11]
```diff
@@ -53,13 +59,29 @@ def treatment_v2(treatment_chemical):
...
+        'award': award['@id'],
+        'lab': lab['@id'],
+        'depletion': False
     }
     return testapp.post_json('/treatment', item, status=201).json['@graph'][0]
+
+
+@pytest.fixture
+def treatment_v3(treatment_chemical):
+    item = treatment_chemical.copy()
```
Comment (iyoungworth) [inline src/igvfd/tests/fixtures/schemas/treatment.py:79]: The fixture should 'pop' depletion property as well to test the upgrade

**Review by iyoungworth — CHANGES_REQUESTED** [review]
Changes requested for title calc and upgrade, I think the dependencies look like they should work

File: src/igvfd/types/treatment.py (hunk 9 above)
Comment (khzlin) [reply src/igvfd/types/treatment.py:41]: no separate statement but changed the text.

File: src/igvfd/upgrade/treatment.py (hunk 10 above)
Comment (khzlin) [reply src/igvfd/upgrade/treatment.py:40]: done

File: src/igvfd/tests/fixtures/schemas/treatment.py (hunk 11 above)
Comment (khzlin) [reply src/igvfd/tests/fixtures/schemas/treatment.py:79]: popped.

**Review by iyoungworth — APPROVED** [review]

**Pull request merged**